### ISSUE TITLE
The question for converting the distance of patch pair to probability using softmax.

### ISSUE THREAD

**Issue by ZhuangPeiyu** [IDLabMedia/comprint#1]
Thanks for sharing this excellent work. However, I have a doubt about the probability conversion. Before calculating the loss, the distance of two patches needs to be converted into probability by the softmax function.
This will make classification probabilities of all patch pairs in the same batch add up to 1.
What is the motivation for doing this? Because I feel that there should be no correlation between each patch pair in one batch. I am looking forward to your reply.

**Comment by hmareen** [comment]
Sorry for my late answer, I did not receive a notification of this issue. Thank you for your insightful comment. I believe that you are correct that there should not be a correlation between each patch pair in one batch, at least not in the current training implementation. However, we believe this is not an issue in this case. I try to explain our reasoning below.

The use of softmax was copied from the [Noiseprint method](https://github.com/grip-unina/noiseprint). This makes sense during training of Noiseprint, as each patch is not just paired with one other patch, but with all patches in the batch. Hence, only one (or perhaps four, as stated in their paper) pair(s) should have have the same fingerprint(s), and all others should not. Then, for each patch, the softmax is calculated by using distances with all other patches that it was paired with in the denominator. Thus, there is some correlation between these pairs, and the softmax makes some sense.

In the current Comprint training method, each patch is paired with only a single other patch. In other words, each patch pair is indeed independent from the other pairs in the batch. The use of softmax seems unintuitive, as it indeed does not make sense to sum the classification probability outputs to 1. However, softmax does help to put each distance in the perspective of all distances in the batch. As such, the interpretation of small and large is given by the normalization that happens during the softmax processing. 

Also, on average, each batch will contain 50% of pairs that should output the same fingerprint (i.e., small distance) and 50% should output a different fingerprint (i.e., large distance). This 50-50 relationship remains approximately constant over the training steps. Thus, the denominator of the softmax max remain approximately constant over the training steps. Additionally, it is important to note that each output should certainly not exceed the value of one. Otherwise, the binary cross entropy loss with the labels applied afterwards becomes invalid.

It could be that there are other solutions than softmax that also put the distances in a correct perspective, while being more intuitive to use. It could certainly be interesting to investigate those.

In conclusion, the output of the softmax should not be interpreted as classification probabilities, but they are only used to put distance values in perspective.
